protocols/autonatv2/src/client/behaviour.rs

```diff
@@ -0,0 +1,228 @@
+        event: <Self::ConnectionHandler as ConnectionHandler>::ToBehaviour,
+    ) {
+        match event {
+            ToBehaviour::ResponseInfo(response_info) => {
+                let event = Event::CompletedReachabilityTest(
+                    self.handle_response_info(peer_id, response_info),
+                );
+                self.pending_events.push_back(ToSwarm::GenerateEvent(event));
+            }
+            _ => todo!(),
+        }
+        todo!()
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<ToSwarm<Self::ToSwarm, <Self::ConnectionHandler as ConnectionHandler>::FromBehaviour>>
+    {
+        if let Some(event) = self.pending_events.pop_front() {
+            return Poll::Ready(event);
+        }
+        if let Some(command) = self.pending_commands.pop_front() {
+            self.handle_command(command);
+            return Poll::Pending;
+        }
+        todo!()
+    }
+}
+
+impl<R> Behaviour<R>
+where
+    R: RngCore,
+{
+    fn handle_response_info(
+        &mut self,
+        peer_id: PeerId,
+        ResponseInfo {
+            response:
+                DialResponse {
+                    status,
+                    dial_status,
+                    ..
+                },
+            suspicious_addrs,
+            successfull_addr,
+        }: ResponseInfo,
+    ) -> Result<ReachabilityTestSucc, ReachabilityTestErr> {
+        match (status, dial_status) {
+            (ResponseStatus::E_REQUEST_REJECTED, _) => {
+                Err(ReachabilityTestErr::ServerRejectedDialRequest { peer_id })
+            }
+            (ResponseStatus::E_DIAL_REFUSED, _) => {
+                Err(ReachabilityTestErr::ServerChoseNotToDialAnyAddress { peer_id })
+            }
+            (ResponseStatus::E_INTERNAL_ERROR, _) => {
+                Err(ReachabilityTestErr::InternalServerError { peer_id })
+            }
+            (ResponseStatus::OK, DialStatus::UNUSED) => {
+                Err(ReachabilityTestErr::InvalidResponse { peer_id })
+            }
+            (ResponseStatus::OK, DialStatus::E_DIAL_ERROR) => {
+                Err(ReachabilityTestErr::UnableToConnectOnSelectedAddress {
+                    peer_id,
+                    addr: successfull_addr,
+                })
+            }
+            (ResponseStatus::OK, DialStatus::E_DIAL_BACK_ERROR) => {
+                Err(ReachabilityTestErr::FailureDuringDialBack {
+                    peer_id,
+                    addr: successfull_addr,
+                })
+            }
+            (ResponseStatus::OK, DialStatus::OK) => Ok(ReachabilityTestSucc {
+                server_peer: peer_id,
+                visible_addr: successfull_addr,
+                suspicious_addrs,
+            }),
+        }
+    }
+
+    fn handle_command(&mut self, command: Command) {
+        match command {
+            Command::TestListenerReachability {
+                server_peer,
+                local_addrs,
```

Comment:
We should gather these addresses from `FromSwarm::NewExternalAddrCandidate`.

Reply:
Done